### DISCUSSION TITLE
Use different text formatting in a single textblock

### DISCUSSION THREAD

**Discussion by AzuxirenLeadGuy** [AvaloniaUI/Avalonia#9654] [Q&A] [answered]
HI. Sorry if this is a silly question, but is there any way to use the same the textblock and have different formatting for different portions of the text?

Somehting Like 

> Lorem **ipsum** dolor sit **amet**, consectetur **adipiscing** elit

Mainly, I want a textblock with certain portions of text bolded/colored. Is it possible to get it in a single textblock?

**Comment by ahopper** [comment, marked as answer]
You certainly can in 11.0, see the bottom of https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/TextBlockPage.xaml

**Reply by AzuxirenLeadGuy** [reply]
This looks great! Thank you for pointing this out.

Now I would also like to set such a style in a Textblock with code (instead of XAML). How can I do that?

**Reply by Gillibald** [reply]
Add inlines to the Inline property

**Reply by AzuxirenLeadGuy** [reply]
Thanks. That was what I needed.

**Reply by AzuxirenLeadGuy** [reply]
@Gillibald and @ahopper Hi just a follow-up question. I have a text block that contains some strings with many inlines. I have used all options of text wrapping, but portions of texts are still not correctly wrapped- Especially at positions where the inlines end/begin. Is this issue known, or am I missing something obvious here?

**Reply by Gillibald** [reply]
Please provide some sample so I can have a proper look

**Reply by AzuxirenLeadGuy** [reply]
@Gillibald Here you go. 

I have created a new `avalonia.xplat` project and went ahead to remove the original `TextBlock` in the MainView within the XAML file. Then I have altered the `MainView.axaml.cs` file as shown:

```
using Avalonia.Controls;
using Avalonia.Controls.Documents;
using Avalonia.Media;

namespace example.Views;

public partial class MainView : UserControl
{
    public const string LongText = "Object initializers let you assign values to any accessible fields or properties of an object at creation time without having to invoke a constructor followed by lines of assignment statements. The object initializer syntax enables you to specify arguments for a constructor or omit the arguments (and parentheses syntax). The following example shows how to use an object initializer with a named type, Cat and how to invoke the parameterless constructor. Note the use of auto-implemented properties in the Cat class. For more information, see Auto-Implemented Properties. The object initializers syntax allows you to create an instance, and after that it assigns the newly created object, with its assigned properties, to the variable in the assignment. Object initializers can set indexers, in addition to assigning fields and properties. Consider this basic Matrix class.";

    public TextBlock block;
    public MainView()
    {
        InitializeComponent();
        block = new TextBlock()
        {
            TextWrapping = Avalonia.Media.TextWrapping.Wrap,
            FontSize = 20,
        };
        Margin = new(5);
        Padding = new(5);

        // Comment one setup at a time

        // Setup1();
        Setup2();

        Content = block;
    }
    public void Setup1()
    {
        block.Text = LongText;
    }
    public void Setup2()
    {
        string[] words = LongText.Split(' ');
        int counter = 5; // Every 5th word is colored
        var inlines = new InlineCollection();
        string sentence = "";
        for(int i=0;i<words.Length;i++)
        {
            if(i%counter==0)
            {
                inlines.Add // Add the normal part of the sentence
                (
                    new TextBlock()
                    {
                        Text = sentence,
                        FontSize = 20,
                        Foreground = new SolidColorBrush(Color.FromRgb(0xff, 0xff, 0xff)), // White color
                        FontStyle = FontStyle.Normal,
                    }
                );
                inlines.Add // Add the stylish word of the sentence
                (
                    new TextBlock()
                    {
                        Text = words[i]+" ",
                        FontSize = 20,
                        Foreground = new SolidColorBrush(Color.FromRgb(0xff, 0xff, 0)), // Yellow color
                        FontStyle = FontStyle.Oblique,
                    }
                );
                sentence = "";
            }
            else
            {
                sentence += $"{words[i]} ";
            }
            block.Inlines = inlines;
        }
    }
}
```

I have used `Setup1()` which puts all text without any formatting, and `Setup2()` which uses formats the text as I need. In the constructor, I run `Setup()` and comment `Setup2()` to obtain 

![image](https://user-images.githubusercontent.com/40014666/206858017-045a3e99-0b22-4f0a-adc3-522a206759cb.png)


But if I use `Setup2()` (and comment out the first setup), then I get this

![image](https://user-images.githubusercontent.com/40014666/206858085-0f41d5a1-7348-497e-8ce8-07aa7c5820c4.png)


As you can see, the text wrapping is not done correctly in this case. Particularly, at the end of the newly added inlines (in yellow color), the text wrapping is incorrect. In this case, the window might be resized and the text looks better, but in my case, I wish to display a lot of texts like this, and simply resizing in such cases is a bit inconvenient.  Please let me know if I'm doing something wrong.

**Reply by Gillibald** [reply]
You are supposed to add Runs and not TextBlocks otherwise you just add implicit InlineUIContainer that wrap a control (TextBlock)

**Reply by AzuxirenLeadGuy** [reply]
@Gillibald Sorry if this is a silly question, but I can't find the type `Runs`. I did find the type `InlineUIContainer` and added objects to the `InlineCollection` while setting its child to the `TextBlock` object I had been adding earlier, but it still looks the same. I did so like this.

```csharp

        string[] words = LongText.Split(' ');
        int counter = 5; // Every 5th word is colored
        var inlines = new InlineCollection();
        string sentence = "";
        for(int i=0;i<words.Length;i++)
        {
            if(i%counter==0)
            {
                inlines.Add // Add the normal part of the sentence
                (
                    new InlineUIContainer()
                    {   
                        Child = new TextBlock()
                        {
                            Text = sentence,
                            FontSize = 20,
                            Foreground = new SolidColorBrush(Color.FromRgb(0xff, 0xff, 0xff)), // White color
                            FontStyle = FontStyle.Normal,                           
                        }
                    }
                );
                inlines.Add // Add the stylish word of the sentence
                (
                    new InlineUIContainer()
                    {
                        Child = new TextBlock()
                        {
                            Text = words[i]+" ",
                            FontSize = 20,
                            Foreground = new SolidColorBrush(Color.FromRgb(0xff, 0xff, 0)), // Yellow color
                            FontStyle = FontStyle.Oblique
                        },
                    }
                );
                sentence = "";
            }
            else
            {
                sentence += $"{words[i]} ";
            }
            block.Inlines = inlines; // block is the TextBlock instance
```

If you don't mind, can you correct my mistake and show me how the code should be?

**Reply by Gillibald** [reply]
https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Controls/Documents

**Comment by Gillibald** [comment]
Replace `InlineUIContainer` with `Run`
